api/sendfild/service/service.go

```diff
@@ -0,0 +1,713 @@
+	}
+	type report struct {
+		All            []entityCount `bson:"all"`
+		ByMessageState []entityCount `bson:"by_message_state"`
+		Waiting        []entityCount `bson:"waiting"`
+		UniqueFrom     []entityCount `bson:"unique_from"`
+		UniqueTo       []entityCount `bson:"unique_to"`
+		SentFilStats   []stats       `bson:"sent_fil_stats"`
+	}
+
+	createdAtMatch := bson.M{}
+	if req.Before != nil {
+		createdAtMatch["$lt"] = req.Before.AsTime()
+	}
+	if req.After != nil {
+		createdAtMatch["$gt"] = req.After.AsTime()
+	}
+	match := bson.M{}
+	if len(createdAtMatch) > 0 {
+		match["created_at"] = createdAtMatch
+	}
+	cursor, err := s.col.Aggregate(ctx, bson.A{
+		bson.M{"$match": match},
+		bson.M{"$facet": bson.M{
+			"all": bson.A{
+				bson.M{"$group": bson.M{"_id": nil, "count": bson.M{"$sum": 1}}},
+			},
+			"by_message_state": bson.A{
+				bson.M{"$group": bson.M{"_id": "$message_state", "count": bson.M{"$sum": 1}}},
+			},
+			"waiting": bson.A{
+				bson.M{"$match": bson.M{"waiting": true}},
+				bson.M{"$group": bson.M{"_id": nil, "count": bson.M{"$sum": 1}}},
+			},
+			"unique_from": bson.A{
+				bson.M{"$group": bson.M{"_id": "$from", "count": bson.M{"$sum": 1}}},
+			},
+			"unique_to": bson.A{
+				bson.M{"$group": bson.M{"_id": "$to", "count": bson.M{"$sum": 1}}},
+			},
+			"sent_fil_stats": bson.A{
+				bson.M{"$group": bson.M{
+					"_id":   nil,
+					"total": bson.M{"$sum": "$amount_nano_fil"},
+					"avg":   bson.M{"$avg": "$amount_nano_fil"},
+					"max":   bson.M{"$max": "$amount_nano_fil"},
+					"min":   bson.M{"$min": "$amount_nano_fil"},
+				}},
+			},
+		}},
+	})
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "calling aggregate: %v", err)
+	}
+	var res []report
+	if err = cursor.All(ctx, &res); err != nil {
+		return nil, status.Errorf(codes.Internal, "decoding cursor results: %v", err)
+	}
+	if len(res) != 1 {
+		return nil, status.Errorf(codes.Internal, "unexpected number of aggregate results: %v", len(res))
+	}
+
+	r := res[0]
+
+	resp := &pb.SummaryResponse{}
+
+	if len(r.All) == 1 {
+		resp.CountTxns = r.All[0].Count
+	}
+
+	for _, state := range r.ByMessageState {
+		switch pb.MessageState(state.ID.(int32)) {
+		case pb.MessageState_MESSAGE_STATE_PENDING:
+			resp.CountPending = state.Count
+		case pb.MessageState_MESSAGE_STATE_ACTIVE:
+			resp.CountActive = state.Count
+		case pb.MessageState_MESSAGE_STATE_FAILED:
+			resp.CountFailed = state.Count
+		}
+	}
+
+	if len(r.Waiting) == 1 {
+		resp.CountWaiting = r.Waiting[0].Count
+	}
+
+	resp.CountFromAddrs = int64(len(r.UniqueFrom))
+	resp.CountToAddrs = int64(len(r.UniqueTo))
+
+	if len(r.SentFilStats) == 1 {
+		resp.TotalNanoFilSent = r.SentFilStats[0].Total
+		resp.AvgNanoFilSent = r.SentFilStats[0].Avg
+		resp.MaxNanoFilSent = r.SentFilStats[0].Max
+		resp.MinNanoFilSent = r.SentFilStats[0].Min
+	}
+
+	return resp, nil
+}
+
+type waitResult struct {
+	txn txn
+	err error
+}
+
+func (s *Service) wait(tx txn) chan waitResult {
```

Comment:
Ok here is where the magic is. 

I love that this func simply returns a chan of `txn` (well, wrapped in a `waitResult
` so we can also pass any error) because it makes the resulting chan easy to consume and the chan output is consistent with the type of data needed in the contexts from which `wait` is called. 

What I don't like is that this func is FULL of side effects. It manages the `waiting` map and lock to keep that in memory state up to date. It communicates with Lotus to wait for the message to become active. It updates the in memory `txn` that is sent on the result chan and updates the underlying mongo representation. This doesn't feel great, but it does keep all the internal state consistent and provide an very clean api, so I went with it 🤷🏼‍♂️

Reply:
Important to note that `wait` is a "get or create" kind of API.

Reply:
This will change in the next PR when I implement waiting and notifying properly. I think the actual waiting will be done in one place as be more of an internal implementation detail. API calls that want to wait will call a subscription type of API and won't have to deal with this currently opaque and poorly named func.